flink-ml-core/src/main/java/org/apache/flink/ml/builder/GraphBuilder.java

```diff
@@ -0,0 +1,431 @@
+     * @param inputs A list of TableIds which represents inputs to fit() of the given Estimator as
+     *     well as inputs to transform() of the Model fitted by the given Estimator.
+     * @return A list of TableIds which represents the outputs of transform() of the Model fitted by
+     *     the given Estimator.
+     */
+    public TableId[] addEstimator(Estimator<?, ?> estimator, TableId... inputs) {
+        return addEstimator(estimator, inputs, inputs);
+    }
+
+    /**
+     * Adds an Estimator in the graph.
+     *
+     * <p>When the graph runs as Estimator, the fit() of the given Estimator would be invoked with
+     * estimatorInputs. Then when the GraphModel fitted by this graph runs, the transform() of the
+     * Model fitted by the given Estimator would be invoked with modelInputs.
+     *
+     * <p>When the graph runs as AlgoOperator or Model, the fit() of the given Estimator would be
+     * invoked with estimatorInputs, then the transform() of the Model fitted by the given Estimator
+     * would be invoked with modelInputs.
+     *
+     * <p>NOTE: the number of the returned TableIds does not represent the actual number of Tables
+     * outputted by transform(). This number could be configured using {@link
+     * #setMaxOutputTableNum(int)}. Users should make sure that this number >= the actual number of
+     * Tables outputted by transform().
+     *
+     * @param estimator An Estimator instance.
+     * @param estimatorInputs A list of TableIds which represents inputs to fit() of the given
+     *     Estimator.
+     * @param modelInputs A list of TableIds which represents inputs to transform() of the Model
+     *     fitted by the given Estimator.
+     * @return A list of TableIds which represents the outputs of transform() of the Model fitted by
+     *     the given Estimator.
+     */
+    public TableId[] addEstimator(
+            Estimator<?, ?> estimator, TableId[] estimatorInputs, TableId[] modelInputs) {
+        return addStage(estimator, StageType.ESTIMATOR, estimatorInputs, modelInputs);
+    }
+
+    /**
+     * When the graph runs as Estimator, it first generates a GraphModel that contains the Model
+     * fitted by the given Estimator. Then when this GraphModel runs, the setModelData() of the
+     * fitted Model would be invoked with the given inputs before its transform() is invoked.
+     *
+     * <p>When the graph runs as AlgoOperator or Model, the setModelData() of the Model fitted by
+     * the given Estimator would be invoked with the given inputs before its transform() is invoked.
+     *
+     * @param estimator An Estimator instance.
+     * @param inputs A list of TableIds which represents inputs to setModelData() of the Model
+     *     fitted by the given Estimator.
+     */
+    public void setModelDataOnEstimator(Estimator<?, ?> estimator, TableId... inputs) {
+        GraphNode node = existingNodes.get(estimator);
+        if (node == null) {
```

Comment:
Could be simplified with `checkState(boolean, "");`, similar to other checks

Reply:
`checkState(...)` throws `IllegalStateException(...)`. According to the this doc [1], `IllegalStateException` signals that a method has been invoked at an illegal or inappropriate time.

In this particular case, we want to throw exception not because `setModelDataOnEstimator(...)` is called, but because the parameter provided to the `setModelDataOnEstimator(...)` is not `compatible` with the state of GraphBuilder. So `IllegalStateException` is probably not a good choice here?

Maybe we can use `checkArgument(...)` which throws `IllegalArgumentException`. My minor concern here is that `IllegalArgumentException` is usually thrown when the argument itself is obviously wrong (e.g. a string-valued config whose value is not recognized). But in this particular case, the exception is thrown based on both the argument and the method's internal state.

I will update the code to use `checkArgument(...)` if you prefer. What do you think?

[1] https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalStateException.html